cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,454 @@
+
+Existing mutual authentication implementations put several restrictions upon users in order to attain additional security guarantees. These can range from requiring apps to use TCP+TLS for authentication/encryption to requiring (sidecar) proxies in the dataplane. These restrictions increase the baseline complexity and processing cost of implementing mutual authentication. We believe that by combining the best of session-based authentication and network-based authentication, we can provide a secure implementation that is faster and more flexible than prior implementations.
+
+## Goals
+
+* Provide a mechanism for pods to opt into peer authentication
+* Pluggable for existing certificate management systems (SPIFFE, Vault, SMI, Istio, cert-manager, …)
+* Configurable certificate granularity
+* Leverage existing encryption protocol support for the dataplane
+* Minimize packets dropped as far as possible in a reasonable amount of effort.
+
+
+## Non-Goals
+
+* Inline authentication via a proxy
+* Guaranteed no-drop connection establishment
+
+
+## Proposal
+
+### Overview
+
+The Isovalent blog [Next-Generation Mutual Authentication with Cilium Service Mesh](https://isovalent.com/blog/post/2022-05-03-servicemesh-security/)
+described the high level design for mutual authentication where there are two channels for connectivity:
+Firstly, a control plane connection between cilium-agent instances on each node provides authentication of connections between pods on the nodes.
+Secondly, the existing Cilium encryption support using WireGuard or IPsec provides an encrypted dataplane for the connections. This document seeks to describe the per-node implementation that coordinates these out-of-band authentication channels with the in-band network encryption channels.
+
+Here is a diagram showing the key parts of the design
+
+![mTLS General Architecture](./images/mTLS-general-architecture.png)
+
+The following sub-sections are broken up by the aspects of the design where specific changes need to be made, starting from the k8s resource types that need to be modified, and followed by the considerations for the cert management control plane, the (local) control plane, and the data plane. While the following sections may focus more on SPIFFE as a default auth implementation for the pluggable interface, the same principles should apply for other auth mechanisms.
+
+### Resources
+
+This is an example of what the config will look like; it's likely that extra fields will need to be added as part of the implementation process.
+
+#### ConfigMap - cilium-config
+
+* option-name: `default`
+  * description
+* mesh-auth-mode: `disabled`
+  * disabled / spiffe / other
+* mesh-auth-spire-server: `____`
+  * Configure the spire server
+* mesh-auth-spire-agent: `unix://var/run/cilium/spiffe/admin/admin.sock`
```

Comment:
Sorry for a late nit: in the PR i named this the `mesh-auth-spire-admin-socket`as the agent itself for workload attestation does not use the admin.sock so while it is the agent i did rename it to be not confused with the usual agent socket

Reply:
Yeah, okay, that's fair, I'll update.